Specifications/QIR/Data-Types.md

```diff
@@ -170,49 +166,36 @@ big integers.
-
-```LLVM
-%TupleHeader = type { i32 }
-```
-
-For instance, a tuple containing two integers, `(Int, Int)`, would be represented in
-LLVM as `type {%TupleHeader, i64, i64}`.
-
-Tuple handles are simple pointers to the underlying data structure.
-When passed to a callable function, tuples should always be passed by reference,
-even if the source language treats them as immutable.
-
-In some situations, a pointer to a tuple of unknown or variable type may be required.
-To satisfy LLVM's strong typing in such a situation, tuple pointers are passed as
-pointers to the initial tuple header field; that is, as a `%TupleHeader*`.
-These pointers should be cast to pointers to the correct data structures by the
+When passed to a callable function, tuples are passed as a pointer to an opaque LLVM structure, `%Tuple`. The pointer is expected to point to the contained data such that it can be cast to the correct data structures by the
 receiving code.
+This permits to define runtime functions that are common for all tuples.
 For instance, this convention is used for callable wrapper functions; see
 [below](#callable-values-and-wrapper-functions).
 
-Many languages provide immutable tuples, along with operators that allow a modified
-copy of an existing tuple to be created.
-In QIR, this is represented by creating a new copy of the existing tuple and then
-modifying the newly-created tuple in place.
-If the compiler knows that the existing tuple is not used after the modification,
-it is possible to avoid the copy and modify the existing tuple in place.
+Many languages provide immutable tuples, along with operators that allow a modified copy of an existing tuple to be created.
+QIR permits to efficiently support this by requiring the runtime to track and be able to access the following given a `%Tuple*`:
+- The size of the tuple in number of bytes
+- The user count indicating how many handles to the tuple exist in the source code
 
-The following utility functions are provided by the classical runtime to support
-tuples and user-defined types:
+The language specific compiler is responsible for injecting calls to increase and decrease the user count as needed, as well as to accurately reflect when references to the LLVM structure representing a tuple are created and removed. 
+See the section [above](#reference-and-access-counting) regarding the distinction between access and reference counting. 
+
+In the case where the source language exposes tuples as value types rather than reference types, the language specific compiler is expected to request the necessary copies prior to modifying the tuple in place. 
+This is done by invoking the runtime function `__quantum__rt__tuple_copy` to create a byte-by-byte copy of a tuple. Unless the copying is forced via the second argument, the runtime may omit copying the value and instead simply return a pointer to the given argument if the user count is 0 and it is hence save to modify the tuple in place.
```

Comment:
```suggestion
This is done by invoking the runtime function `__quantum__rt__tuple_copy` to create a byte-by-byte copy of a tuple. Unless the copying is forced via the second argument, the runtime may omit copying the value and instead simply return a pointer to the given argument if the user count is 0 and it is therefore safe to modify the tuple in place.
```

Also, why do we need a "forced" copy? For that case, why wouldn't the compiler just generate the proper sequence of loads and stores (and user/ref count updates for inner pointers)?

Reply:
There will be legitimate cases when a client might want to copy even in absence of aliases, no?

Reply:
Yes, a compiler may choose not to track alias counts and still need the option to have a copy be made.